hera_mc/node.py

```diff
@@ -0,0 +1,259 @@
+# -*- mode: python; coding: utf-8 -*-
+# Copyright 2018 the HERA Collaboration
+# Licensed under the 2-clause BSD license.
+
+"""Node M&C info from the node's Redis database
+
+"""
+from __future__ import absolute_import, division, print_function
+
+from astropy.time import Time
+from math import floor
+from sqlalchemy import Column, BigInteger, Integer, Float, Boolean
+
+from . import MCDeclarativeBase
+
+defaultServerAddress = 'redishost'
+
+sensor_key_dict = {'top_sensor_temp': 'temp_top', 'middle_sensor_temp': 'temp_mid',
+                   'bottom_sensor_temp': 'temp_bot', 'humidity_sensor_temp': 'temp_humid',
+                   'humidity': 'humid'}
+
+power_status_key_dict = {'snap_relay_powered': 'power_snap_relay', 'snap0_powered': 'power_snap_0',
+                         'snap1_powered': 'power_snap_1', 'snap2_powered': 'power_snap_2',
+                         'snap3_powered': 'power_snap_3', 'pam_powered': 'power_pam',
+                         'fem_powered': 'power_fem'}
+
+
+class NodeSensor(MCDeclarativeBase):
+    """
+    Definition of node sensor table.
+
+    time: gps time of the node data, floored (BigInteger, part of primary_key).
+    node: node number (Integer, part of primary_key)
+    top_sensor_temp: temperature of top sensor reported by node in Celcius
+    middle_sensor_temp: temperature of middle sensor reported by node in Celcius
+    bottom_sensor_temp: temperature of bottom sensor reported by node in Celcius
+    humidity_sensor_temp: temperature of the humidity sensor reported by node in Celcius
+    humidity: percent humidity measurement reported by node
+    """
+    __tablename__ = 'node_sensor'
+    time = Column(BigInteger, primary_key=True)
+    node = Column(Integer, primary_key=True)
+    top_sensor_temp = Column(Float)
```

Comment:
Looking at the Alembic migration script, it looks like all of these fields will be nullable. Would it be better if they were non-nullable? I feel like it's generally good to start strict and loosen up if needed ... but on the other hand, I have trouble imagining realistic scenarios where nullability gets us in trouble. Maybe if we're getting bad data from the sensors, which end up getting represented as Nones, and then we don't notice that we're not taking any data for a while.

Reply:
So it sounds like other than the bottom temperature sensor, there will always be some value reported for all the sensors. But if the sensor goes bad or can't be reached the reported value will be `-99`. I'm trying to decide if it's more helpful to have `-99`s in this table representing bad readings or `None`s.

Reply:
Ok, so these columns are going to stay nullable (per Slack conversation) because the sensors can give bad readings. Those readings will be reported to this code as `None`s. The columns in the `node_power_status` table will be made not nullable, however, because we don't expect to have missing readings for those.